moonraker/plugins/history.py

```diff
@@ -0,0 +1,253 @@
+# History cache for printer jobs
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+import json, logging, time
+from tornado.ioloop import IOLoop
+
+SAVE_INTERVAL = 5
+
+HIST_NAMESPACE = "history"
+METADATA_KEYS = ['estimated_time','modified','slicer','slicer_version']
+
+class History:
+    def __init__(self, config):
+        self.server = config.get_server()
+        self.database = self.server.lookup_plugin("database")
+        self.gcdb = self.database.wrap_namespace("gcode_metadata",
+            parse_keys=False)
+        self.current_job = None
+        self.file_metadata = None
+        self.jobs = {}
+        self.job_id = -1
+        self.last_update_time = 0
+        self.print_stats = {}
+
+        self.server.register_event_handler(
+            "server:klippy_ready", self._init_ready)
+        self.server.register_event_handler(
+            "server:status_update", self._status_update)
+
+
+        self.server.register_endpoint(
+            "/printer/history/list", ['GET'], self._handle_jobs_list)
+        self.server.register_endpoint(
+            "/printer/history/delete", ['DELETE'], self._handle_job_delete)
+
+        self.database.register_local_namespace(HIST_NAMESPACE)
+        try:
+            self.job_id = self.database.get_item(HIST_NAMESPACE,
+                "job_auto_inc_id")
+        except:
+            logging.info("Creating job history namespace in database")
+            self.database.insert_item(HIST_NAMESPACE,"job_auto_inc_id", 0)
+            self.job_id = 0
+            self.database.insert_item(HIST_NAMESPACE,"prints", self.jobs)
+
+        jobs = self.database.get_item(HIST_NAMESPACE,"prints")
+        for job in jobs:
+            self.jobs[job] = PrinterJob(jobs[job])
+
+    async def _init_ready(self):
+        klippy_apis = self.server.lookup_plugin('klippy_apis')
+        self.print_stats = {}
+
+        try:
+            result = await klippy_apis.query_objects({'print_stats': None})
+        except self.server.error as e:
+            logging.info(f"Error getting print_stats: {e}")
+        self.print_stats = result.get("print_stats", {})
+
+        sub = {"print_stats": None}
+        try:
+            status = await klippy_apis.subscribe_objects(sub)
+        except self.server.error as e:
+            logging.info(f"Error subscribing to print_stats")
+
+    async def _handle_job_delete(self, web_request):
+        args = web_request.get_args()
+        if "all" in args:
```

Comment:
Rather than using argument keys as parameters I think it would be better to specify named arguments.  For example, the endpoint could look like:
```
DELETE /server/history/delete?job=jobname
```
or to clear all:
```
DELETE /server/history/delete?all=true
```
Then the handler could do something like
```python
clear_all = web_request.get_boolean("all", False)
if clear_all:
  # delete all jobs
else:
  jobname = web_request.get_str("job")
  # delete associated job
```